UPGRADE_GUIDE.md

```diff
@@ -2,7 +2,203 @@
+- [Upgrading from 2.x to 3.0](#upgrading-from-2x-to-30)
+
+## Upgrading from 3.x to 4.0
+
+### 4.0 High Impact Changes
+
+- [Updating Dependencies](#40-updating-dependencies)
+- [Client Instance](#40-client-instance)
+- [Language Conventions](#40-language-conventions)
+- [Nullable Properties](#40-nullable-properties)
+
+### 4.0 Medium Impact Changes
+
+- [Error Types](#40-error-types)
+- [Beta Client](#40-beta-client)
+
+### 4.0 Low Impact Changes
+
+- [Services and Resources](#40-services-and-resources)
+
+## 4.0 Updating Dependencies
+
+*Likelihood of Impact: **High***
+
+**Dependencies**
+
+- RestSharp was upgraded from v107 to v108
+- All dependencies had minor version bumps
+
+## 4.0 Client Instance
+
+*Likelihood of Impact: **High***
+
+The library is now designed around the idea of a `Client`. Users will initialize a `Client` instance with their API key
+and then use that instance to make API calls.
+
+This change allows for multiple clients to be instantiated with different API keys and for the library to be used in a
+multi-threaded environment.
+
+```csharp
+// Old
+EasyPost.ClientManager.SetCurrent("EASYPOST_API_KEY"); // global API key
+
+// New
+var client1 = new EasyPost.Client("EASYPOST_API_KEY_1"); // per-client API key
+var client2 = new EasyPost.Client("EASYPOST_API_KEY_2"); // per-client API key
+```
+
+All functions are now accessed via the `Client` instance. For example, to create a shipment:
+
+```csharp
+// Old
+EasyPost.Shipment shipment = await EasyPost.Shipment.Create(parameters);
+
+// New
+EasyPost.Shipment shipment = awiat myClient.Shipment.Create(parameters);
+```
+
+## 4.0 Language Conventions
+
+*Likelihood of Impact: **High***
+
+All functions, classes and properties have been renamed to follow the C# language conventions. The most notable changes
+here is that all properties of an object are now PascalCase rather than snake_case.
+
+```csharp
+// Old
+var id = myAddress.id;
+var toAddress = myShipment.to_address;
+var formType = myScanForm.form_file_type;
+
+// New
+var id = myAddress.Id;
+var toAddress = myShipment.ToAddress;
+var formType = myScanForm.FormFileType;
+```
+
+## 4.0 Nullable Properties
+
+*Likelihood of Impact: **High***
+
+All properties of all objects are now nullable. This means that if a property is not returned by the API, it will be
+either `null` or the default value for the given type (e.g. `false` for `bool`).
+
+This change silences compilation warnings for the library.
+
+The majority of properties for an object will likely be populated as a result of an API call, but if not, end-users will
+know that the value of a non-populated property is guaranteed to be `null`.
+
+End users may need to add additional null checks or nullability markers in their code, although this change will likely
+play well with any existing null checks.
+
+```csharp
+// Old
+string id = myAddress.id; // "id" is non-nullable, but not guaranteed to be populated
```

Comment:
Issue: ID is always guaranteed to be populated heh. Let's choose something that is not guaranteed to be nullable so the example is more appropriate. This as written is confusing.

Reply:
Resolved